File: trunk/src/cgi/msg_cgi/cgi_msg_system_message_get.cpp
#include "cgi_msg_server.h"
#include "proto_io_tcp_client.h"
#include <time.h>
#include "msg.pb.h"

class CgiMsgSystemMessageGet: public CgiMsgServer
{
public:
	CgiMsgSystemMessageGet():
		CgiMsgServer(0, "config.ini", "logger.properties")
	{

	}

	bool InnerProcess()
	{
		//1.get HTTP params
		uint64_t openid_md5 = strtoul(((string)GetInput().GetValue("openid_md5")).c_str(), NULL, 10);
		uint64_t pa_appid_md5 = strtoul(((string)GetInput().GetValue("pa_appid_md5")).c_str(), NULL, 10);
		uint64_t begin_ts = strtoul(((string)GetInput().GetValue("begin_ts")).c_str(), NULL, 10);
		uint64_t limit = strtoul(((string)GetInput().GetValue("limit")).c_str(), NULL, 10);

		if(0 == openid_md5 || 0 == pa_appid_md5 || (limit <=0 || limit >= 100))
		{
			DoReply(CGI_RET_CODE_INVALID_PARAM);
			LOG4CPLUS_ERROR(logger, "INVALID openid_md5="<<openid_md5<<", pa_appid_md5="<<pa_appid_md5<<", begin_ts="<<begin_ts<<", limit="<<limit);
			return true;	
		}		
				
		//2.Build requestMsg protobuf
		std::string strErrMsg = "";
		::hoosho::msg::Msg stRequestMsg;
		::hoosho::msg::Msg stResponseMsg;
		
		::hoosho::msg::MsgHead* stRequestMsgHead = stRequestMsg.mutable_head();
		stRequestMsgHead->set_cmd(::hoosho::msg::QUERY_SYSTEM_MSG_REQ);		
		stRequestMsgHead->set_seq(time(NULL));

		::hoosho::sixin::QuerySystemMsgReq* stMsgQueryReq = stRequestMsg.mutable_system_msg_query_req();
		stMsgQueryReq->set_openid_md5(openid_md5);
		stMsgQueryReq->set_pa_appid_md5(pa_appid_md5);
		stMsgQueryReq->set_begin_ts(begin_ts);
		stMsgQueryReq->set_limit(limit);

		//3.send to server, and recv responseMsg protobuf
		common::protoio::ProtoIOTcpClient ioclient(msgServerIP, msgServerPort);

		int iRet;
		iRet = ioclient.io(stRequestMsg, stResponseMsg, strErrMsg);

		if(iRet != 0)
		{
			DoReply(CGI_RET_CODE_SERVER_BUSY);
			LOG4CPLUS_ERROR(logger, "ProtoIOTcpClient IO failed, errmsg = " << strErrMsg);
			return true;
		}

		LOG4CPLUS_DEBUG(logger, "response Msg: \n" << stResponseMsg.Utf8DebugString());	
		//parse response
		const ::hoosho::msg::MsgHead& stHead = stResponseMsg.head();
		if(stHead.cmd() != ::hoosho::msg::QUERY_SYSTEM_MSG_RES)
		{
			DoReply(CGI_RET_CODE_SERVER_BUSY);
			LOG4CPLUS_ERROR(logger, "response.cmd="<<stHead.cmd()<<", unknown, fuck!!!");
			return true;	
		}

		if(stHead.result() != ::hoosho::msg::E_OK)
		{
			DoReply(CGI_RET_CODE_SERVER_BUSY);
			LOG4CPLUS_ERROR(logger, "response.result="<<stHead.result());
			return true;	
		}

		//4.Build strResponse Json from responseMsg protobuf			
		const ::hoosho::sixin::QuerySystemMsgRes& stMsgQueryRes = stResponseMsg.system_msg_query_res();
		for(int i=0; i<stMsgQueryRes.system_msg_list_size(); i++)
		{
			lce::cgi::CAnyValue stAnyValue;
			stAnyValue["pa_appid_md5"] = stMsgQueryRes.system_msg_list(i).pa_appid_md5();
			stAnyValue["create_ts"] = stMsgQueryRes.system_msg_list(i).create_ts();
			stAnyValue["title"] = stMsgQueryRes.system_msg_list(i).title();
			stAnyValue["content"] = stMsgQueryRes.system_msg_list(i).content();
			GetAnyValue()["system_msg_list"].push_back(stAnyValue);
		}
		
		DoReply(CGI_RET_CODE_OK);
		return true;		
	}
	
};

int main()
{
	CgiMsgSystemMessageGet cgi;
	if(!cgi.Run())
	{
		return -1;
	}
	return 0;
}

File: trunk/src/cgi_python/msg_cgi/cgi_msg_system_message_get.py
#!/usr/bin/python
# -*- coding: UTF-8 -*-

import sys
import time

# 包含上级目录，即../common
sys.path.append('..')

# cgi, proto_io
from common.cgi.cgi_ret_code_def import CGI_RET_CODE
import common.cgi.cgi_my as _cgi
import common.proto_io.proto_io_tcp_client as _tcp_client

# proto
import common.proto.msg_pb2 as _msg
import common.proto.msg_sixin_pb2 as _msg_sixin
import common.proto.comm_struct_pb2 as _comm_struct
import common.proto.comm_enum_pb2 as _comm_enum

# util
from common.util.common_util import *

def InnerProcess():
	# 1.Get HTTP params
	openid_md5 = int(mycgi.getEnvValue('openid_md5', 0))
	pa_appid_md5 = int(mycgi.getEnvValue('pa_appid_md5', 0))
	begin_ts = int(mycgi.getEnvValue('begin_ts', 0))
	limit = int(mycgi.getEnvValue('limit', 0))

	# 2.Params check
	if 0 == openid_md5 or 0 == pa_appid_md5 or (limit <= 0 or limit >= 100):
		logger.error('INVALID PARAM: openid_md5=%d, pa_appid_md5=%d, limit=%d !', openid_md5, pa_appid_md5, limit)
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_INVALID_PARAM)		
		return True

	# 3.Build requestMsg protobuf
	stRequestMsg = _msg.Msg()

	stRequestMsgHeader = stRequestMsg.head
	stRequestMsgHeader.cmd = _msg.QUERY_SYSTEM_MSG_REQ
	stRequestMsgHeader.seq = int(time.time())

	stRequestMsgBody = stRequestMsg.system_msg_query_req
	stRequestMsgBody.openid_md5 = openid_md5
	stRequestMsgBody.pa_appid_md5 = pa_appid_md5
	stRequestMsgBody.begin_ts = begin_ts
	stRequestMsgBody.limit = limit

	logger.debug('requestMsg: \n %s', stRequestMsg)

	# 4.Send to server, and recv responseMsg protobuf	
	ioclient = _tcp_client.ProtoIOTcpClient(msgServerIP, msgServerPort)

	dictMsg = {'requestMsg':stRequestMsg, 'responseMsg':None, 'strErrMsg':None}

	ret = ioclient.io(dictMsg)

	if ret != 0:	
		logger.error('ProtoIOTcpClient IO failed, errmsg = %s', dictMsg['strErrMsg'])
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_SERVER_BUSY)		
		return True

	stResponseMsg = dictMsg['responseMsg']
	logger.debug('responseMsg: \n %s', stResponseMsg)

	if stResponseMsg.head.cmd != _msg.QUERY_SYSTEM_MSG_RES:
		logger.error('response.head.cmd=%d, unkown, fuck!!!', stResponseMsg.head.cmd)
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_SERVER_BUSY)		
		return True

	if stResponseMsg.head.result != _msg.E_OK:
		logger.error('response.head.result=%d, not E_OK!', stResponseMsg.head.result)
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_SERVER_BUSY)		
		return True

	# 5.protobuf -> json, and do reply
	stMsgQueryRes = stResponseMsg.system_msg_query_res
	system_msg_list = []
	for systemMsg in stMsgQueryRes.system_msg_list:
		dictItem = {}
		dictItem['pa_appid_md5'] = systemMsg.pa_appid_md5
		dictItem['create_ts'] = systemMsg.create_ts
		dictItem['title'] = systemMsg.title
		dictItem['content'] = systemMsg.content
		system_msg_list.append(dictItem)

	mycgi.addBody('system_msg_list', system_msg_list)
	mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_OK)
	return True





if __name__ == '__main__':
	# __file__ 获取模块所在路径
	mycgi = _cgi.myCgi(__file__, 'logging.conf')

	# logger	
	logger = mycgi.getLogger()

	InnerProcess()	